packages/next/src/server/lib/dedupe-fetch.ts

```diff
@@ -74,29 +83,35 @@ export function createDedupeFetch(originalFetch: typeof fetch) {
     }
 
     const cacheEntries = getCacheEntries(url)
-    let match
-    if (cacheEntries.length === 0) {
-      // We pass the original arguments here in case normalizing the Request
-      // doesn't include all the options in this environment.
-      match = originalFetch(resource, options)
-      cacheEntries.push(cacheKey, match)
-    } else {
-      // We use an array as the inner data structure since it's lighter and
-      // we typically only expect to see one or two entries here.
-      for (let i = 0, l = cacheEntries.length; i < l; i += 2) {
-        const key = cacheEntries[i]
-        const value = cacheEntries[i + 1]
-        if (key === cacheKey) {
-          match = value
-          // I would've preferred a labelled break but lint says no.
-          return match.then((response: Response) => response.clone())
-        }
-      }
-      match = originalFetch(resource, options)
-      cacheEntries.push(cacheKey, match)
+
+    // Check if there is a cached entry for the given cache key. If there is, we
+    // return the cached response (cloned). This will keep the cached promise to
+    // remain unused and can be cloned on future requests.
+    let entry = cacheEntries[cacheKey]
+    if (entry) {
+      return entry.promise.then(() => {
+        if (!entry.response) throw new InvariantError('No cached response')
+
+        const [cloned1, cloned2] = cloneResponse(entry.response)
```

Comment:
I really don't love the `tee` API and having to implement `cloneResponse` in a similar fashion is unfortunate. I think we should just stash the original body in a third slot in the cache entries array and tee it there. That way cloneResponse doesn't have to leak into the patch-fetch implementation. This assumes you keep the array impl but if you are comitted to using an object then just stash the body in another property of the entry

Reply:
As this identified a problem with the `Response#clone()` implementation, I'd assume that we'd want to address it in both the `patch-fetch` and `dedupe-fetch` wrappers.